<h1>Activity #1: Palmer Penguin Data Analytics Project</h1>
<hr>

<h4>Analyst: Karl Gabriel Dellova</h4>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
from scipy.stats import f_oneway
import matplotlib.pyplot as plt

In [2]:
pingu = pd.read_csv('data_sets\\penguins.csv')

In [3]:
pingu

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>Dis Descriptive Analytics</h1>
<sh>Just Testing</sh>

In [4]:
pingu.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [5]:
pingu.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

In [6]:
pingu['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [7]:
pingu['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

<h1>Dis for Adelies</h1>

In [8]:
pingu['species'] == "Adelie"

0       True
1       True
2       True
3       True
4       True
       ...  
339    False
340    False
341    False
342    False
343    False
Name: species, Length: 344, dtype: bool

In [9]:
pingu[pingu['species'] == "Adelie"]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

<h1>Dis for Gentoo</h1>

pingu['species'] == "Gentoo"

In [10]:
pingu[pingu['species'] == "Gentoo"]

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [11]:
pingu['species'] == "Chinstrap"

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: species, Length: 344, dtype: bool

In [12]:
pingu[pingu['species'] == "Chinstrap"]

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278       3650.0    male  2007  
279       3525.0  female  2007  
280       3725.0    male  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[68 rows x 8 columns]

In [13]:
adelie_df = pingu[pingu['species'] == "Adelie"]
gentoo_df = pingu[pingu['species'] == "Gentoo"]
chinstrap_df = pingu[pingu['species'] == "Chinstrap"]

In [14]:
adelie_df['body_mass_g'].mean()

np.float64(3700.662251655629)

<h1>Multiple Conditionals</h1>

In [15]:
thik_pingu = pingu[(pingu['species'] == "Chinstrap") & (pingu['body_mass_g'] >= 3500)]

In [16]:
thik_pingu


species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276  Chinstrap  Dream            46.5           17.9              192.0   
277  Chinstrap  Dream            50.0           19.5              196.0   
278  Chinstrap  Dream            51.3           19.2              193.0   
279  Chinstrap  Dream            45.4           18.7              188.0   
280  Chinstrap  Dream            52.7           19.8              197.0   
281  Chinstrap  Dream            45.2           17.8              198.0   
283  Chinstrap  Dream            51.3           18.2              197.0   
284  Chinstrap  Dream            46.0           18.9              195.0   
285  Chinstrap  Dream            51.3           19.9              198.0   
286  Chinstrap  Dream            46.6           17.8              193.0   
287  Chinstrap  Dream            51.7           20.3              194.0   
288  Chinstrap  Dream            47.0           17.3              185.0   
289  Chinstrap  Dream            52.0           18.1              201.0   
290  Chinstrap  Dream            45.9           17.1              190.0   
291  Chinstrap  Dream            50.5           19.6              201.0   
293  Chinstrap  Dream            58.0           17.8              181.0   
295  Chinstrap  Dream            49.2           18.2              195.0   
296  Chinstrap  Dream            42.4           17.3              181.0   
299  Chinstrap  Dream            50.6           19.4              193.0   
301  Chinstrap  Dream            52.0           19.0              197.0   
303  Chinstrap  Dream            49.5           19.0              200.0   
304  Chinstrap  Dream            46.4           17.8              191.0   
305  Chinstrap  Dream            52.8           20.0              205.0   
307  Chinstrap  Dream            54.2           20.8              201.0   
309  Chinstrap  Dream            51.0           18.8              203.0   
310  Chinstrap  Dream            49.7           18.6              195.0   
311  Chinstrap  Dream            47.5           16.8              199.0   
312  Chinstrap  Dream            47.6           18.3              195.0   
313  Chinstrap  Dream            52.0           20.7              210.0   
315  Chinstrap  Dream            53.5           19.9              205.0   
316  Chinstrap  Dream            49.0           19.5              210.0   
317  Chinstrap  Dream            46.2           17.5              187.0   
318  Chinstrap  Dream            50.9           19.1              196.0   
319  Chinstrap  Dream            45.5           17.0              196.0   
320  Chinstrap  Dream            50.9           17.9              196.0   
321  Chinstrap  Dream            50.8           18.5              201.0   
323  Chinstrap  Dream            49.0           19.6              212.0   
325  Chinstrap  Dream            49.8           17.3              198.0   
327  Chinstrap  Dream            51.4           19.0              201.0   
328  Chinstrap  Dream            45.7           17.3              193.0   
329  Chinstrap  Dream            50.7           19.7              203.0   
333  Chinstrap  Dream            49.3           19.9              203.0   
334  Chinstrap  Dream            50.2           18.8              202.0   
335  Chinstrap  Dream            45.6           19.4              194.0   
336  Chinstrap  Dream            51.9           19.5              206.0   
337  Chinstrap  Dream            46.8           16.5              189.0   
338  Chinstrap  Dream            45.7           17.0              195.0   
339  Chinstrap  Dream            55.8           19.8              207.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
276       3500.0  female  2007  
277       3900.0    male  2007  
278    

<h1>Cleaning the data</h1>

In [17]:
pingu

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

In [18]:
pingu['bill_length_mm'].mean()
pingu['bill_depth_mm'].mean()
pingu['flipper_length_mm'].mean()
pingu['body_mass_g'].mean()
pingu['sex'].mode()

0    male
Name: sex, dtype: object

In [19]:
pingu['bill_length_mm'] = pingu['bill_length_mm'].fillna(pingu['bill_length_mm'].mean())
pingu['bill_depth_mm'] = pingu['bill_depth_mm'].fillna(pingu['bill_depth_mm'].mean())
pingu['flipper_length_mm'] = pingu['flipper_length_mm'].fillna(pingu['flipper_length_mm'].mean())
pingu['body_mass_g'] = pingu['body_mass_g'].fillna(pingu['body_mass_g'].mean())
pingu['sex'] = pingu['sex'].fillna(pingu['sex'].mode()[0])

In [20]:
pingu.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [21]:
pingu['species'] = pingu['species'].astype('category')
pingu['island'] = pingu['island'].astype('category')
pingu['sex'] = pingu['sex'].astype('category')

In [22]:
pingu['body_mass_g'] = pingu['body_mass_g'].astype('float64')

In [23]:
pingu['bill_length_mm'].mean()
pingu['bill_depth_mm'].mean()
pingu['flipper_length_mm'].mean()
pingu['sex'].mode()

0    male
Name: sex, dtype: category
Categories (2, object): ['female', 'male']

In [24]:
adelie_df['body_mass_g'].mean()

np.float64(3700.662251655629)

In [25]:
gentoo_df['body_mass_g'].mean()

np.float64(5076.016260162602)

In [26]:
chinstrap_df['body_mass_g'].mean()

np.float64(3733.0882352941176)

<h1>Q1: Which penguin has the biggest flipper in each island?</h1>

In [27]:
pingu['island'].value_counts()

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

In [28]:
pingu['flipper_length_mm'].mean()

np.float64(200.91520467836258)

In [29]:
big_flipper = pingu.groupby(['island','species'])[['flipper_length_mm']].mean()
big_flipper

flipper_length_mm
island    species                     
Biscoe    Adelie            188.795455
          Chinstrap                NaN
          Gentoo            217.055768
Dream     Adelie            189.732143
          Chinstrap         195.823529
          Gentoo                   NaN
Torgersen Adelie            191.382985
          Chinstrap                NaN
          Gentoo                   NaN

<h4>Insight.1: Gentoo has the longest flipper in Biscoe, Chinstrap on Dream island, and Adelie on Torgersen</h4>
<hr>
<hr>

<h1>Q2. Whats the max bill lenght of each penguin species</h1>
<hr>

In [30]:
pingu.groupby('species')[['bill_length_mm']].mean()

bill_length_mm
species                  
Adelie          38.825144
Chinstrap       48.833824
Gentoo          47.475983

<h4>Insight.2: Gentoo has the longest bill among the three species of penguin. Chinstrap having the longest bill length</h4>
<hr>
<hr>

<h1>Q3. What is the distribution of sex on each island?</h1>
<hr>

In [31]:
pingu[['sex', 'island', 'species']].groupby(['sex', 'island']).count()

species
sex    island            
female Biscoe          80
       Dream           61
       Torgersen       24
male   Biscoe          88
       Dream           63
       Torgersen       28

<h4>Insight.3: Biscoe has the most amount of male and female penguins among the three. Torgersen on the otherhand has the least amount of both male and female penguins among the three</h4>
<hr>
<hr>

<h1>Q4. How many of the three penguin species were distributed within (2007–2009)?</h1>
<hr>

In [32]:
pingu['year'].value_counts()

year
2009    120
2008    114
2007    110
Name: count, dtype: int64

In [33]:
pingu.groupby(['year', 'species']).size().reset_index(name='count')

year    species  count
0  2007     Adelie     50
1  2007  Chinstrap     26
2  2007     Gentoo     34
3  2008     Adelie     50
4  2008  Chinstrap     18
5  2008     Gentoo     46
6  2009     Adelie     52
7  2009  Chinstrap     24
8  2009     Gentoo     44

<h4>Insight.4: Adelies are the most ditributed penguins in 2007 to 2009 with the highest amount being 52. Chinstrap on the otherhand was the least distributed penguin in 2007 to 2009 with the highest amount being 26</h4>
<hr>
<hr>

<h1>Q5. What year had the highest recorded average bill lenght across all species?</h1>
<hr>

In [34]:
pingu.groupby(['year', 'species'])['bill_length_mm'].mean().reset_index()

year    species  bill_length_mm
0  2007     Adelie       38.926439
1  2007  Chinstrap       48.723077
2  2007     Gentoo       47.014706
3  2008     Adelie       38.560000
4  2008  Chinstrap       48.700000
5  2008     Gentoo       46.936957
6  2009     Adelie       38.982692
7  2009  Chinstrap       49.054167
8  2009     Gentoo       48.395953

<h4>Insight.5: The year with the highest recorded average bill lenght was during 2009 belonging to the chinstrap species with the lowest belonging to the adelie species during 2008</h4>
<hr>
<hr>

<h1>Q6. Is there a significant change in average flipper length of penguins across the years 2007–2009?</h1>
<hr>

In [35]:
pingu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(3), float64(4), int64(1)
memory usage: 15.0 KB


In [36]:
pingu_07 = pingu[pingu['year'] == 2007]
pingu_08 = pingu[pingu['year'] == 2008]
pingu_09 = pingu[pingu['year'] == 2009]
f_oneway(pingu_07['flipper_length_mm'], pingu_08['flipper_length_mm'], pingu_09['flipper_length_mm'])

F_onewayResult(statistic=np.float64(6.795857507110622), pvalue=np.float64(0.0012761391137243068))

<h1>Q7. What is the average flipper length of each species of penguin, in each island by sex?</h1>
<hr>

In [37]:
pingu.groupby(['species', 'island', 'sex'])['flipper_length_mm'].mean()

species    island     sex   
Adelie     Biscoe     female    187.181818
                      male      190.409091
           Dream      female    187.851852
                      male      191.482759
           Torgersen  female    188.291667
                      male      194.032686
Chinstrap  Biscoe     female           NaN
                      male             NaN
           Dream      female    191.735294
                      male      199.911765
           Torgersen  female           NaN
                      male             NaN
Gentoo     Biscoe     female    212.706897
                      male      220.877503
           Dream      female           NaN
                      male             NaN
           Torgersen  female           NaN
                      male             NaN
Name: flipper_length_mm, dtype: float64

<h4>Insight.7: The species with the longest average flipper length belong to a male gentoo penguin on biscoe island with a flpper lenght of 220mm</h4>
<hr>
<hr>

<h1>Q8. What's the max flipper length and body mass of each species of penguin by sex</h1>

In [38]:
pingu.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(3), float64(4), int64(1)
memory usage: 15.0 KB


In [39]:
max_pingu = pingu.groupby(['species', 'sex'])[['flipper_length_mm', 'body_mass_g']].max()
max_pingu

flipper_length_mm  body_mass_g
species   sex                                   
Adelie    female              202.0       3900.0
          male                210.0       4775.0
Chinstrap female              202.0       4150.0
          male                212.0       4800.0
Gentoo    female              222.0       5200.0
          male                231.0       6300.0

<h4>Insight.8: The penguin species with the hightes max recorded flipper lenght and body mass is both the male and female Gentoo penguins</h4>
<hr>
<hr>

<h1>Q9. What's the max bill length of each species of female penguin</h1>

In [40]:
f_pingu = pingu[pingu['sex'] == 'female']

In [41]:
f_bill = f_pingu.groupby(['island', 'species'])['bill_length_mm'].max()
f_bill

island     species  
Biscoe     Adelie       40.5
           Chinstrap     NaN
           Gentoo       50.5
Dream      Adelie       42.2
           Chinstrap    58.0
           Gentoo        NaN
Torgersen  Adelie       41.1
           Chinstrap     NaN
           Gentoo        NaN
Name: bill_length_mm, dtype: float64

<h4>Insight.9: The max bill length of a female penguin is 50.5 belonging to the gentoo species residing in biscoe island</h4>
<hr>
<hr>

<h1>Q10. What's the max recorded weight of female gentoo's during 2009</h1>

In [42]:
max_female_09 = f_pingu[(f_pingu['species']=='Gentoo') & (f_pingu['year'] == 2009)]['body_mass_g'].max()
max_female_09

5200.0

<h4>Insight.10: It shows the max body mass of a female Gentoo penguin in 2009 is 5200g</h4>
<hr>
<hr>

<h1>Q11. Which penguin species has the smallest variation in flipper_length?</h1>

In [43]:
pingu_weight_variation = pingu.groupby("species")['flipper_length_mm'].std()
pingu_weight_variation

species
Adelie       6.578130
Chinstrap    7.131894
Gentoo       6.621801
Name: flipper_length_mm, dtype: float64

<h4>Insigth.11: This show the smallest variation in flipper length is 6.57</h4>
<hr>
<hr>

<h1>Q12. Which penguin species has the smallest recorded bill depth in Biscoe during 2008</h1>

In [44]:
pingu_08 = pingu['year'] == '2008'

In [47]:
# lill_bill = pingu_08.groupby(['island', 'species'])['bill_depth_mm'].min()
# lill_bill
#(Testing lng po to)

In [ ]:
lil_bill = pingu[(pingu['island'] == 'Biscoe') & (pingu['year'] == 2009)]['bill_depth_mm'].min()
lil_bill

<h1>Data Aggregation</h1>

In [ ]:
spe_is_y = pingu.groupby(['island','species', 'sex'])['flipper_length_mm'].max()

In [ ]:
spe_is_y

In [ ]:
bill_le_cor = pingu[(adelie_df['flipper_length_mm'].corr(adelie_df['bill_length_mm']))]
bill_le_cor

In [ ]:
island_df_biscoe = pingu[pingu['island'] == 'biscoe']['flipper_length_mm']
island_df_biscoe